Fix #6982: Client side localization script and context-param

melloware opened the pull request:
- Refactored HeadRender to have an `encodeJS` function and reuse it for validation, bv-validation, and now locales.js.
- Adds new context-param `PRIME_ICONS` which is `true` by default but allows someone to disable PrimeIcons. At my company we use a custom theme that is all FontAwesome 5.12 icons so we don't need this CSS loaded.
- Adds new context-param `CLIENT_SIDE_LOCALISATION` which is `false` by default. If enabled will add the `locales-#{language.code}.js` file to the page.

```xml
    <context-param>
        <param-name>primefaces.CLIENT_SIDE_LOCALISATION</param-name>
        <param-value>true</param-value>
    </context-param>
```

@jepsar please review the Dutch translation against the US English on the Wiki and provide me an update before this PR is merged if you want the Dutch updated!

melloware commented:
Showcase PR: https://github.com/primefaces/primefaces-showcase/pull/323

tandraschko commented:
1) we should use a better config name. Maybe primefaces.PRIME_ICONS and primefaces.CLIENT_SIDE_LOCALISATION
2) we should add only the script for the current locale (also its better to maintain in separated files)

jepsar commented:
I will update and extend the Dutch version. As @tandraschko already mentioned, please load the scripts per language. I was planning to load them in the order `locale.getLanguage() + "_" + locale.getCountry()` and then (if not found) `locale.getLanguage()`. If you want me to create files, let met know.

melloware commented:
Good ideas thanks for the feedback!

Rapster commented on src/main/resources/META-INF/resources/primefaces/locales/locales.js:
```diff
@@ -0,0 +1,1363 @@
+    prevText: "Précédent",
+    nextText: "Suivant",
+    monthNames: ["Janvier", "Février", "Mars", "Avril", "Mai", "Juin", "Juillet", "Août", "Septembre", "Octobre", "Novembre", "Décembre"],
+    monthNamesShort: ["Jan", "Fév", "Mar", "Avr", "Mai", "Jun", "Jul", "Aoû", "Sep", "Oct", "Nov", "Déc"],
```
A few typos here: https://fr.wikipedia.org/wiki/Mois#Abr%C3%A9viations (i guess it's always been there)

melloware commented:
OK made all the changes requested from @tandraschko and @jepsar .  I broke the files just down into language like `locale-en.js` not separate `locale-en_US.js` files. I load the file by its language and since our client side code already handles trying to load the more specific one en_US then fallback to `en` I thought this made the Language files more manageable.   For example `locale-pt.js` has both Portuguese and Brazilian Portugeuse in it.

@Rapster feel free to edit `locale-fr.js` right in this PR if you want. Just use the GitHub online editor.

jepsar commented on docs/10_0_0/gettingstarted/configuration.md:
```diff
@@ -14,6 +14,7 @@ pattern of JavaEE. Here is the list of all configuration options defined with a
 | Name | Default | Description |
 | --- | --- | --- |
 | CLIENT_SIDE_VALIDATION | false | Enables/disables global client side validation . |
+| CLIENT_SIDE_LOCALISATION | false | Adds locale.js automatically for your locale. |
```
```suggestion
| CLIENT_SIDE_LOCALISATION | false | Adds `"locales/locale-" + locale.getLanguage() + ".js"` automatically for your locale. |
```

jepsar commented on src/main/java/org/primefaces/renderkit/HeadRenderer.java:
```diff
@@ -123,6 +127,11 @@ public void encodeBegin(FacesContext context, UIComponent component) throws IOEx
 
+        if (applicationContext.getConfig().isClientSideLocalizationEnabled()) {
+            Locale locale = LocaleUtils.getCurrentLocale(context);
+            encodeJS(context, "locales/locale-" + locale.getLanguage() + ".js");
```
Should the user make sure that all their application locale files are present? I would say so. Something to point out in the documentation.

melloware replied:
I actually just updated to throw an exception. If you are using CLIENT_SIDE_LOCALISATION and your file does not exist we should throw a FacesException like the encodeCSS function does if the CSS is missing.

jepsar commented on src/main/resources/META-INF/resources/primefaces/locales/locale-fa.js:
```diff
@@ -0,0 +1,18 @@
+    dayNamesShort: ['یک', 'دو', 'سه', 'چهار', 'پنج', 'جمعه', 'شنبه'],
+    dayNamesMin: ['ی', 'د', 'س', 'چ', 'پ', 'ج', 'ش'],
+    weekHeader: 'ه',
+    dateFormat: 'yy/mm/dd',
```
This is the only locale with a `dateFormat`.

tandraschko commented:
i just wonder a bit about lang vs locale here.
What happens for portugese vs brazil portugese? both lang are "pt" but different locale?

melloware commented:
So What happens is this....

We load `locale-pt.js` which has both. But when The user sends their code to the UI it will be "pt-BR" or "pt-PT".   That is why in the `locale-en.js` file I have...
```javascript
 PrimeFaces.locales['en_US'] = PrimeFaces.locales['en'];
 PrimeFaces.locales['en_GB'] = PrimeFaces.locales['en']; 
```
I think we should always declare the major language like `fr` and then have mappings to other like `fr_CA` that map to FR unless someone does have specifics for that country like we do with Portuguese. 

Thoughts?

I can add more language codes in each one if we want. For example for French I could add...
```
  "fr-BE",
  "fr-CA",
  "fr-CH",
  "fr-FR",
  "fr-LU",
  "fr-MC",
```

tandraschko commented:
yep, i see
looks good to me

melloware commented:
OK working on it now.

tandraschko commented:
i mean that the current status of the PR is ok for me :D

jepsar commented:
> Thoughts?
> 
> I can add more language codes in each one if we want. For example for French I could add...

People can allways overload the resource file if they need customizations... or report them.

That was the reason I got triggered by the date format being only available for one locale. If it would be present for all, I would have added it for `nl` and put a different one for `nl_BE`.

jepsar commented on src/main/resources/META-INF/resources/primefaces/locales/locale-nl.js:
```diff
@@ -0,0 +1,102 @@
+        'javax.validation.constraints.Pattern.message': 'moet overeenkomen met "{0}".',
+        'javax.validation.constraints.Size.message': 'grootte moet tussen {0} en {1} zijn.'
+    }
+ };
```
```suggestion
 };
 PrimeFaces.locales['nl_BE'] = PrimeFaces.locales['nl'];
```

melloware commented:
OK added all specific locales per file.

melloware commented:
I think this will give us a good starting point too for people to report any errors or updates as they see them.   I think the English file is always the one that other languages should be compared against.

tandraschko commented:
i dont think we should add this:
```PrimeFaces.locales['fr-BE'] = PrimeFaces.locales['fr'];```

our clients side has a fallback to fr, if fr-BE isnt found. right?

jepsar commented:
Why are `-` and `_` mixed? Eg `en_US` vs `fr-BE`.

Rapster commented:
Ideally, if no client locale-file defined then it should fallback just like @melloware defined. For example, if no fr-BE then PF should fallback on fr. Not sure how easy this is since our js files are loaded first

melloware commented:
@Rapster I think my latest changes will definitely accomplish that.  `fr_BE` is mapped by default to `fr`  If however someone down the road says its wrong and provides us with a more specific `fr_BE` we simply swap out `PrimeFaces.locales['fr-BE'] = PrimeFaces.locales['fr'];` for their new specific mapping.

tandraschko commented:
hm, i dont understand it.
If you check core.js getLocaleSettings, we already have a fallback, so this fallback:
```PrimeFaces.locales['fr-BE'] = PrimeFaces.locales['fr'];```
in the locale files isnt required.

melloware commented:
@jepsar I think you are right they should all be underscores as PF send "en_US" not "en-US".

@tandraschko It looks like that is not used everywhere and for example `clock.js` just has `this.cfg.monthNames = PrimeFaces.locales[this.cfg.locale].monthNames;`  it looks like that getLocaleSettings was only used in validation.js not everywhere.  So I thought it safer to add the specific mappings just in case anyone else has written code not using the `getLocaleSettings` method.

tandraschko commented:
i think we should use getLocaleSettings everything, everything else is more boilerplate code.

melloware commented:
Agreed want me to make that change everywhere? Looks like its used in 5 widgets directly using the locale.

tandraschko commented:
yep please

melloware commented:
Removed all extra language mappings except `en_US`. 
Refactored all spots to now use `PrimeFaces.getLocaleSettings()`
 `PrimeFaces.getLocaleSettings()` is now smarter by allowing an optional locale from the widget.
Removed custom getLocaleSettings from the bean validation.js and switch that to use the one from core

tandraschko commented:
looks good for me
we just need to wait for the OK of PrimeTek.

jepsar commented:
Great job! Thank you.

melloware commented:
I have to make one more fix. The settings shouldn't be cached if passing the widget config locale.

melloware commented:
OK updated and tested.  Now it only caches the global settings once per page but when asking for a widget config it looks it up every time.  This is because like our DatePicker example two widgets on the same page can be different languages.

melloware commented:
I need to update our MD doc for Localization next; https://primefaces.github.io/primefaces/10_0_0/#/core/localization?id=localization

buunsal commented:
```
/** Turkish */
PrimeFaces.locales['tr'] = {
    closeText: 'Kapat',
    prevText: 'Geri',
    nextText: 'İleri',
    monthNames: ['Ocak', 'Şubat', 'Mart', 'Nisan', 'Mayıs', 'Haziran', 'Temmuz', 'Ağustos', 'Eylül', 'Ekim', 'Kasım', 'Aralık'],
    monthNamesShort: ['Oca', 'Şub', 'Mar', 'Nis', 'May', 'Haz', 'Tem', 'Ağu', 'Eyl', 'Eki', 'Kas', 'Ara'],
    dayNames: ['Pazar', 'Pazartesi', 'Salı', 'Çarşamba', 'Perşembe', 'Cuma', 'Cumartesi'],
    dayNamesShort: ['Pz', 'Pt', 'Sa', 'Ça', 'Pe', 'Cu', 'Ct'],
    dayNamesMin: ['Pz', 'Pt', 'Sa', 'Ça', 'Pe', 'Cu', 'Ct'],
    weekHeader: 'Hf',
    weekNumberTitle: 'Hf',
    firstDay: 1,
    isRTL: false,
    showMonthAfterYear: false,
    yearSuffix: '',
    timeOnlyTitle: 'Zaman Seçiniz',
    timeText: 'Zaman',
    hourText: 'Saat',
    minuteText: 'Dakika',
    secondText: 'Saniye',
    currentText: 'Geçerli Tarih',
    ampm: false,
    year: 'Yıl',
    month: 'Ay',
    week: 'Hafta',
    day: 'Gün',
    list: 'Agenda',
    allDayText: 'Tüm Gün',
    moreLinkText: 'Daha Fazla...',
    noEventsText: 'Etkinlik Yok',
    today: 'Bugün',
    clear: 'Temizle',
    aria: {
        'paginator.PAGE': 'Sayfa {0}',
        'calendar.BUTTON': 'Takvimi Göster',
        'datatable.sort.ASC': 'Sütunu artan sıralamak için etkinleştir',
        'datatable.sort.DESC': 'Sütunu azalan sıralamak için etkinleştir',
        'columntoggler.CLOSE': 'Kapat',
        'overlaypanel.CLOSE': 'Kapat'
    },
    messages: {  //optional for Client Side Validation
        'javax.faces.component.UIInput.REQUIRED': '{0}: Doğrulama Hatası: Değer gerekli.',
        'javax.faces.converter.IntegerConverter.INTEGER': '{2}: \'{0}\' sayısı bir veya daha fazla rakamdan oluşmalıdır.',
        'javax.faces.converter.IntegerConverter.INTEGER_detail': '{2}: \'{0}\' sayısı -2147483648 ve 2147483647 arasında olmalıdır. Örnek: {1}.',
        'javax.faces.converter.DoubleConverter.DOUBLE': '{2}: \'{0}\' sayısı bir veya daha fazla rakamdan oluşmalıdır.',
        'javax.faces.converter.DoubleConverter.DOUBLE_detail': '{2}: \'{0}\' sayısı 4.9E-324 ve 1.7976931348623157E308 arasında olmalıdır. Örnek: {1}.',
        'javax.faces.converter.BigDecimalConverter.DECIMAL': '{2}: \'{0}\' işaretli bir ondalık sayı olmalıdır.',
        'javax.faces.converter.BigDecimalConverter.DECIMAL_detail': '{2}: \'{0}\' sıfır veya daha fazla basamaktan oluşan işaretli bir ondalık sayı olmalıdır, ardından bir ondalık virgül veya kesir gelebilir. Örnek: {1}.',
        'javax.faces.converter.BigIntegerConverter.BIGINTEGER': '{2}: \'{0}\' sayısı bir veya daha fazla rakamdan oluşmalıdır.',
        'javax.faces.converter.BigIntegerConverter.BIGINTEGER_detail': '{2}: \'{0}\' sayısı bir veya daha fazla rakamdan oluşmalıdır. Örnek: {1}.',
        'javax.faces.converter.ByteConverter.BYTE': '{2}: \'{0}\' sayısı 0 ve 255 arasında olmalıdır.',
        'javax.faces.converter.ByteConverter.BYTE_detail': '{2}: \'{0}\' sayısı 0 ve 255 arasında olmalıdır. Örnek: {1}.',
        'javax.faces.converter.CharacterConverter.CHARACTER': '{1}: \'{0}\' değeri geçerli bir karakter olmalı.',
        'javax.faces.converter.CharacterConverter.CHARACTER_detail': '{1}: \'{0}\' değeri geçerli bir ASCII karakteri olmalı.',
        'javax.faces.converter.ShortConverter.SHORT': '{2}: \'{0}\' sayısı bir veya daha fazla rakamdan oluşmalıdır.',
        'javax.faces.converter.ShortConverter.SHORT_detail': '{2}: \'{0}\' sayısı -32768 ve 32767 arasında olmalıdır. Örnek: {1}.',
        'javax.faces.converter.BooleanConverter.BOOLEAN': '{1}: \'{0}\' değeri \'true\' (doğru) yada \'false\' (yanlış) olmalıdır.',
        'javax.faces.converter.BooleanConverter.BOOLEAN_detail': '{1}: \'{0}\' değeri \'true\' (doğru) yada \'false\' (yanlış) olmalıdır. \'true\' dışındaki değerler \'false\' olarak değerlendirilir.',
        'javax.faces.validator.LongRangeValidator.MAXIMUM': '{1}: Doğrulama Hatası: Değer izin verilen maksimum değerden daha büyük: \'{0}\'.',
        'javax.faces.validator.LongRangeValidator.MINIMUM': '{1}: Doğrulama Hatası: Değer izin verilen minimum değerden daha düşük: \'{0}\'.',
        'javax.faces.validator.LongRangeValidator.NOT_IN_RANGE': '{2}: Doğrulama Hatası: Belirtilen öznitelik {0} ve {1} değerleri arasında değil.',
        'javax.faces.validator.LongRangeValidator.TYPE={0}': 'Doğrulama Hatası: Değer doğru türde değil.',
        'javax.faces.validator.DoubleRangeValidator.MAXIMUM': '{1}: Doğrulama Hatası: Değer izin verilen maksimum değerden daha büyük: \'{0}\'.',
        'javax.faces.validator.DoubleRangeValidator.MINIMUM': '{1}: Doğrulama Hatası: Değer izin verilen minimum değerden daha düşük: \'{0}\'.',
        'javax.faces.validator.DoubleRangeValidator.NOT_IN_RANGE': '{2}: Doğrulama Hatası: Belirtilen öznitelik {0} ve {1} değerleri arasında değil.',
        'javax.faces.validator.DoubleRangeValidator.TYPE={0}': 'Doğrulama Hatası: Değer doğru türde değil.',
        'javax.faces.converter.FloatConverter.FLOAT': '{2}: \'{0}\' sayısı bir veya daha fazla rakamdan oluşmalıdır.',
        'javax.faces.converter.FloatConverter.FLOAT_detail': '{2}: \'{0}\' sayısı 1.4E-45 ve 3.4028235E38 arasında olmalıdır. Örnek: {1}.',
        'javax.faces.converter.DateTimeConverter.DATE': '{2}: \'{0}\' değeri tarih olarak anlaşılamadı.',
        'javax.faces.converter.DateTimeConverter.DATE_detail': '{2}: \'{0}\' değeri  tarih olarak anlaşılamadı. Örnek: {1}.',
        'javax.faces.converter.DateTimeConverter.TIME': '{2}: \'{0}\' değeri  saat olarak anlaşılamadı.',
        'javax.faces.converter.DateTimeConverter.TIME_detail': '{2}: \'{0}\' değeri  saat olarak anlaşılamadı. Örnek: {1}.',
        'javax.faces.converter.DateTimeConverter.DATETIME': '{2}: \'{0}\' değeri tarih ve saat olarak anlaşılamadı.',
        'javax.faces.converter.DateTimeConverter.DATETIME_detail': '{2}: \'{0}\' değeri tarih ve saat olarak anlaşılamadı. Örnek: {1}.',
        'javax.faces.converter.DateTimeConverter.PATTERN_TYPE': '{1}: \'{0}\' değerini dönüştürmek için \'pattern\' yada \'type\' değeri girilmelidir.',
        'javax.faces.converter.NumberConverter.CURRENCY': '{2}: \'{0}\' değeri parasal birim olarak anlaşılamadı.',
        'javax.faces.converter.NumberConverter.CURRENCY_detail': '{2}: \'{0}\' değeri parasal birim olarak anlaşılamadı. Örnek: {1}.',
        'javax.faces.converter.NumberConverter.PERCENT': '{2}: \'{0}\' değeri yüzdesel olarak anlaşılamadı.',
        'javax.faces.converter.NumberConverter.PERCENT_detail': '{2}: \'{0}\' değeri yüzdesel olarak anlaşılamadı. Örnek: {1}.',
        'javax.faces.converter.NumberConverter.NUMBER': '{2}: \'{0}\' değeri sayı olarak anlaşılamadı.',
        'javax.faces.converter.NumberConverter.NUMBER_detail': '{2}: \'{0}\' değeri sayı olarak anlaşılamadı. Örnek: {1}.',
        'javax.faces.converter.NumberConverter.PATTERN': '{2}: \'{0}\' değeri sayı deseni olarak anlaşılamadı.',
        'javax.faces.converter.NumberConverter.PATTERN_detail': '{2}: \'{0}\' değeri sayı deseni olarak anlaşılamadı. Örnek: {1}.',
        'javax.faces.validator.LengthValidator.MINIMUM': '{1}: Doğrulama Hatası: Uzunluk, izin verilen minimum değerden daha az: \'{0}\'.',
        'javax.faces.validator.LengthValidator.MAXIMUM': '{1}: Doğrulama Hatası: Uzunluk izin verilen maksimum değerden daha büyük: \'{0}\'.',
        'javax.faces.validator.RegexValidator.PATTERN_NOT_SET': 'Regex deseni boş olamaz.',
        'javax.faces.validator.RegexValidator.PATTERN_NOT_SET_detail': 'Regex deseni boş değer alamaz.',
        'javax.faces.validator.RegexValidator.NOT_MATCHED': 'Regex deseni eşleşmedi.',
        'javax.faces.validator.RegexValidator.NOT_MATCHED_detail': '\'{0}\' ile regex deseni eşleşmedi.',
        'javax.faces.validator.RegexValidator.MATCH_EXCEPTION': 'Regex hatası.',
        'javax.faces.validator.RegexValidator.MATCH_EXCEPTION_detail': 'Regex hatası, \'{0}\'.',
        'javax.faces.validator.BeanValidator.MESSAGE': '{0}',
        'javax.validation.constraints.AssertFalse.message': 'Değer \'false\' (yanlış) olmak zorunda.',
        'javax.validation.constraints.AssertTrue.message': 'Değer \'true\' (doğru) olmak zorunda.',
        'javax.validation.constraints.DecimalMax.message': '{0} veya daha düşük olmalı.',
        'javax.validation.constraints.DecimalMin.message': '{0} veya daha yüksek olmalı.',
        'javax.validation.constraints.Digits.message': 'Sayısal değer sınırların dışında (&lt;{0} basamak&gt;.&lt;{1} basamak&gt; olmalı).',
        'javax.validation.constraints.Future.message': 'Gelecek bir tarih olmalı.',
        'javax.validation.constraints.Max.message': '{0} veya daha düşük olmalı.',
        'javax.validation.constraints.Min.message': '{0} veya daha yüksek olmalı',
        'javax.validation.constraints.NotNull.message': '\'null\' (boş) olmayabilir.',
        'javax.validation.constraints.Null.message': '\'null\' (boş) olmalıdır.',
        'javax.validation.constraints.Past.message': 'Geçmiş bir tarih olmalıdır.',
        'javax.validation.constraints.Pattern.message': '"{0}" ile eşleşmelidir.',
        'javax.validation.constraints.Size.message': '{0} ile {1} arasında bir değer olmalıdır.'
    }
};
```

jepsar commented on src/main/resources/META-INF/resources/primefaces/locales/locale-it.js:
```diff
@@ -0,0 +1,27 @@
+    week: 'Settimana',
+    day: 'Giorno',
+    allDayText: 'Tutto il Giorno'
+};
```
```suggestion
/** Italian */
PrimeFaces.locales ['it'] = {
    closeText: 'Chiudi',
    prevText: 'Precedente',
    nextText: 'Successivo',
    monthNames: ['Gennaio', 'Febbraio', 'Marzo', 'Aprile', 'Maggio', 'Giugno', 'Luglio', 'Agosto', 'Settembre', 'Ottobre', 'Novembre', 'Dicembre'],
    monthNamesShort: ['Gen', 'Feb', 'Mar', 'Apr', 'Mag', 'Giu', 'Lug', 'Ago', 'Set', 'Ott', 'Nov', 'Dic'],
    dayNames: ['Domenica', 'Lunedì', 'Martedì', 'Mercoledì', 'Giovedì', 'Venerdì', 'Sabato'],
    dayNamesShort: ['Dom', 'Lun', 'Mar', 'Mer', 'Gio', 'Ven', 'Sab'],
    dayNamesMin: ['D', 'L', 'M', 'M ', 'G', 'V ', 'S'],
    weekHeader: 'Sett',
    weekNumberTitle: 'S',
    firstDay: 1,
    isRTL: false,
    showMonthAfterYear: false,
    yearSuffix: '',
    timeOnlyTitle: 'Solo orario',
    timeText: 'Orario',
    hourText: 'Ore',
    minuteText: 'Minuti',
    secondText: 'Secondi',
    currentText: 'Oggi',
    ampm: false,
    year: 'Anno',
    month: 'Mese',
    week: 'Settimana',
    day: 'Giorno',
    list: 'Agenda',
    allDayText: 'Tutto il giorno',
    moreLinkText: 'Altro...',
    noEventsText: 'Nessun evento',
    today: 'Oggi',
    clear: 'Pulisci',
    aria: {
        'paginator.PAGE': 'Pagina {0}',
        'calendar.BUTTON': 'Mostra calendario',
        'datatable.sort.ASC': 'attiva per ordinare la colonna in senso crescente',
        'datatable.sort.DESC': 'attiva per ordinare la colonna in senso descrescente',
        'columntoggler.CLOSE': 'Chiudi',
        'overlaypanel.CLOSE': 'Chiudi'
    },
    messages: {  //optional for Client Side Validation
        'javax.faces.component.UIInput.REQUIRED': '{0}: Errore di validazione: Valore obbligatorio.',
        'javax.faces.converter.IntegerConverter.INTEGER': '{2}: \'{0}\' dev\'essere un numero costituito da uno o più cifre.',
        'javax.faces.converter.IntegerConverter.INTEGER_detail': '{2}: \'{0}\' dev\'essere un numero compreso tra -2147483648 e 2147483647. Esempio: {1}.',
        'javax.faces.converter.DoubleConverter.DOUBLE': '{2}: \'{0}\' dev\'essere un numero costituito da uno o più cifre.',
        'javax.faces.converter.DoubleConverter.DOUBLE_detail': '{2}: \'{0}\' dev\'essere un numero compreso tra 4.9E-324 e 1.7976931348623157E308. Esempio: {1}.',
        'javax.faces.converter.BigDecimalConverter.DECIMAL': '{2}: \'{0}\' dev\'essere un numero decimale con segno.',
        'javax.faces.converter.BigDecimalConverter.DECIMAL_detail': '{2}: \'{0}\' dev\'essere un numero decimale con segno costituito da zero o più cifre, eventualmente seguito dalla virgola e da una parte frazionaria. Esempio: {1}.',
        'javax.faces.converter.BigIntegerConverter.BIGINTEGER': '{2}: \'{0}\' dev\'essere un numero costituito da una o più cifre.',
        'javax.faces.converter.BigIntegerConverter.BIGINTEGER_detail': '{2}: \'{0}\' dev\'essere un numero costituito da una o più cifre. Esempio: {1}.',
        'javax.faces.converter.ByteConverter.BYTE': '{2}: \'{0}\' dev\'essere un numero compreso tra 0 e 255.',
        'javax.faces.converter.ByteConverter.BYTE_detail': '{2}: \'{0}\' dev\'essere un numero compreso tra 0 e 255. Esempio: {1}.',
        'javax.faces.converter.CharacterConverter.CHARACTER': '{1}: \'{0}\' dev\'essere un carattere valido.',
        'javax.faces.converter.CharacterConverter.CHARACTER_detail': '{1}: \'{0}\' dev\'essere un carattere ASCII.',
        'javax.faces.converter.ShortConverter.SHORT': '{2}: \'{0}\' dev\'essere un numero costituito da una o più cifre.',
        'javax.faces.converter.ShortConverter.SHORT_detail': '{2}: \'{0}\' dev\'essere un numero compreso tra -32768 e 32767. Esempio: {1}.',
        'javax.faces.converter.BooleanConverter.BOOLEAN': '{1}: \'{0}\' dev\'essere \'vero\' o \'falso\'.',
        'javax.faces.converter.BooleanConverter.BOOLEAN_detail': '{1}: \'{0}\' dev\'essere \'vero\' o \'falso\'. Qualunque altro valore diverso da \'vero\' verrà valutato come \'falso\'.',
        'javax.faces.validator.LongRangeValidator.MAXIMUM': '{1}: Errore di validazione: Il valore è maggiore del massimo valore consentito di \'{0}\'.',
        'javax.faces.validator.LongRangeValidator.MINIMUM': '{1}: Errore di validazione: Il valore è minore del minimo valore consentito di \'{0}\'.',
        'javax.faces.validator.LongRangeValidator.NOT_IN_RANGE': '{2}: Errore di validazione: Il valore specificato non è compreso nell\'intervallo previsto tra {0} e {1}.',
        'javax.faces.validator.LongRangeValidator.TYPE={0}': 'Errore di validazione: Il valore non è del tipo corretto.',
        'javax.faces.validator.DoubleRangeValidator.MAXIMUM': '{1}: Errore di validazione: Il valore è maggiore del massimo consentito di \'{0}\'.',
        'javax.faces.validator.DoubleRangeValidator.MINIMUM': '{1}: Errore di validazione: Il valore è minore del minimo consentito di \'{0}\'.',
        'javax.faces.validator.DoubleRangeValidator.NOT_IN_RANGE': '{2}: Errore di validazione: Il valore specificato non è compreso nell\'intervallo previsto tra {0} e {1}.',
        'javax.faces.validator.DoubleRangeValidator.TYPE={0}': 'Errore di validazione: Il valore non è del tipo corretto.',
        'javax.faces.converter.FloatConverter.FLOAT': '{2}: \'{0}\' dev\'essere un numero costituito da una o più cifre.',
        'javax.faces.converter.FloatConverter.FLOAT_detail': '{2}: \'{0}\' dev\'essere un numero compreso tra 1.4E-45 e 3.4028235E38. Esempio: {1}.',
        'javax.faces.converter.DateTimeConverter.DATE': '{2}: non è stato possibile interpretare \'{0}\' come una data.',
        'javax.faces.converter.DateTimeConverter.DATE_detail': '{2}: non è stato possibile interpretare \'{0}\' come una data. Esempio: {1}.',
        'javax.faces.converter.DateTimeConverter.TIME': '{2}: non è stato possibile interpretare \'{0}\' come un orario.',
        'javax.faces.converter.DateTimeConverter.TIME_detail': '{2}: non è stato possibile interpretare \'{0}\' come un orario. Esempio: {1}.',
        'javax.faces.converter.DateTimeConverter.DATETIME': '{2}: non è stato possibile interpretare \'{0}\' come una data ed un orario.',
        'javax.faces.converter.DateTimeConverter.DATETIME_detail': '{2}: non è stato possibile interpretare \'{0}\' come una data ed un orario. Esempio: {1}.',
        'javax.faces.converter.DateTimeConverter.PATTERN_TYPE': '{1}: Almeno uno degli attributi \'pattern\' o \'type\' dev\'essere specificato per convertire il valore \'{0}\'.',
        'javax.faces.converter.NumberConverter.CURRENCY': '{2}: non è stato possibile interpretare \'{0}\' come un valore monetario.',
        'javax.faces.converter.NumberConverter.CURRENCY_detail': '{2}: non è stato possibile interpretare \'{0}\' come un valore monetario. Esempio: {1}.',
        'javax.faces.converter.NumberConverter.PERCENT': '{2}: non è stato possibile interpretare \'{0}\' come una percentuale.',
        'javax.faces.converter.NumberConverter.PERCENT_detail': '{2}: non è stato possibile interpretare \'{0}\' come una percentuale. Esempio: {1}.',
        'javax.faces.converter.NumberConverter.NUMBER': '{2}: non è stato possibile interpretare \'{0}\' come un numero.',
        'javax.faces.converter.NumberConverter.NUMBER_detail': '{2}: non è stato possibile interpretare \'{0}\' come un numero. Esempio: {1}.',
        'javax.faces.converter.NumberConverter.PATTERN': '{2}: non è stato possibile interpretare \'{0}\' come un pattern numerico.',
        'javax.faces.converter.NumberConverter.PATTERN_detail': '{2}: non è stato possibile interpretare \'{0}\' come un pattern numerico. Esempio: {1}.',
        'javax.faces.validator.LengthValidator.MINIMUM': '{1}: Errore di validazione: La lunghezza è minore del minimo consentito di \'{0}\'.',
        'javax.faces.validator.LengthValidator.MAXIMUM': '{1}: Validation Error: La lunghezza è maggiore del massimo consentito di \'{0}\'.',
        'javax.faces.validator.RegexValidator.PATTERN_NOT_SET': 'L\'espressione regolare dev\'essere impostata.',
        'javax.faces.validator.RegexValidator.PATTERN_NOT_SET_detail': 'Il pattern di espressione regolare dev\'essere impostato ad un valore non vuoto.',
        'javax.faces.validator.RegexValidator.NOT_MATCHED': 'Mancata corrispondenza dell\'espressione regolare.',
        'javax.faces.validator.RegexValidator.NOT_MATCHED_detail': 'L\'espressione regolare \'{0}\' non è stata soddisfatta.',
        'javax.faces.validator.RegexValidator.MATCH_EXCEPTION': 'Errore nell\'espressione regolare.',
        'javax.faces.validator.RegexValidator.MATCH_EXCEPTION_detail': 'Errore nell\'espressione regolare, \'{0}\'.',
        //optional for bean validation integration in client side validation
        'javax.faces.validator.BeanValidator.MESSAGE': '{0}',
        'javax.validation.constraints.AssertFalse.message': 'dev\'essere falso.',
        'javax.validation.constraints.AssertTrue.message': 'dev\'essere vero.',
        'javax.validation.constraints.DecimalMax.message': 'dev\'essere minore o uguale a {0}.',
        'javax.validation.constraints.DecimalMin.message': 'dev\'essere maggiore o uguale a {0}.',
        'javax.validation.constraints.Digits.message': 'valore numerico fuori intervallo (numero di cifre previste compreso tra {0} e {1}).',
        'javax.validation.constraints.Future.message': 'dev\'essere nel futuro.',
        'javax.validation.constraints.Max.message': 'dev\'essere minore o uguale a {0}.',
        'javax.validation.constraints.Min.message': 'dev\'essere maggiore o uguale a {0}.',
        'javax.validation.constraints.NotNull.message': 'non può essere nullo.',
        'javax.validation.constraints.Null.message': 'dev\'essere nullo.',
        'javax.validation.constraints.Past.message': 'dev\'essere nel passato.',
        'javax.validation.constraints.Pattern.message': 'deve soddisfare l\'espressione "{0}".',
        'javax.validation.constraints.Size.message': 'la dimensione dev\'essere compresa tra {0} e {1}.'
    }
};
```

stolp commented:
Let me work at the German translation.

stolp commented:
German:

I built this based on the code from the Wiki, there were quite a few changes necessary. hope I catched them all. I am also updating this in the Wiki.

```
PrimeFaces.locales ['de'] = {
    closeText: 'Schließen',
    prevText: 'Zurück',
    nextText: 'Weiter',
    monthNames: ['Januar', 'Februar', 'März', 'April', 'Mai', 'Juni', 'Juli', 'August', 'September', 'Oktober', 'November', 'Dezember' ],
    monthNamesShort: ['Jan', 'Feb', 'Mär', 'Apr', 'Mai', 'Jun', 'Jul', 'Aug', 'Sep', 'Okt', 'Nov', 'Dez' ],
    dayNames: ['Sonntag', 'Montag', 'Dienstag', 'Mittwoch', 'Donnerstag', 'Freitag', 'Samstag'],
    dayNamesShort: ['Son', 'Mon', 'Die', 'Mit', 'Don', 'Fre', 'Sam'],
    dayNamesMin: ['S', 'M', 'D', 'M ', 'D', 'F ', 'S'],
    weekHeader: 'Woche',
    weekNumberTitle: 'W',
    firstDay: 1,
    isRTL: false,
    showMonthAfterYear: false,
    yearSuffix:'',
    timeOnlyTitle: 'Nur Zeit',
    timeText: 'Zeit',
    hourText: 'Stunde',
    minuteText: 'Minute',
    secondText: 'Sekunde',
    currentText: 'Aktuelles Datum',
    ampm: false,
    year: 'Jahr',
    month: 'Monat',
    week: 'Woche',
    day: 'Tag',
    list: 'Agenda',
    allDayText: 'Ganzer Tag',
    moreLinkText: 'Mehr...',
    noEventsText: 'Keine Einträge',
    today: 'Heute',
    clear: 'Löschen',
    aria: {
        'paginator.PAGE': 'Seite {0}',
        'calendar.BUTTON': 'Kalender anzeigen',
        'datatable.sort.ASC': 'Spalten aufsteigend sortieren',
        'datatable.sort.DESC': 'Spalten absteigend sortieren',
        'columntoggler.CLOSE': 'Schließen',
        'overlaypanel.CLOSE': 'Schließen'
    },
    messages: {  //optional for Client Side Validation
        'javax.faces.component.UIInput.REQUIRED': '{0}: Eingabe erforderlich.',
        'javax.faces.converter.IntegerConverter.INTEGER': '{2}: \'{0}\' muss eine Zahl mit einer oder mehr Ziffern sein.',
        'javax.faces.converter.IntegerConverter.INTEGER_detail': '{2}: \'{0}\' muss eine Zahl zwischen -2147483648 und 2147483647 sein. Beispiel: {1}',
        'javax.faces.converter.DoubleConverter.DOUBLE': '{2}: \'{0}\' muss eine Zahl mit einer oder mehr Ziffern sein.',
        'javax.faces.converter.DoubleConverter.DOUBLE_detail': '{2}: \'{0}\' muss eine Zahl zwischen 4.9E-324 und 7976931348623157E308 sein. Beispiel: {1}',
        'javax.faces.converter.BigDecimalConverter.DECIMAL': '{2}: \'{0}\' muss eine Dezimalzahl mit Vorzeichen sein.',
        'javax.faces.converter.BigDecimalConverter.DECIMAL_detail': '{2}: \'{0}\' muss eine Dezimalzahl mit Vorzeichen sein, die aus Null oder mehr Ziffern besteht, gefolgt von Dezimalstellen nach einem Punkt. Beispiel: {1}',
        'javax.faces.converter.BigIntegerConverter.BIGINTEGER': '{2}: \'{0}\' muss eine Zahl mit einer oder mehr Ziffern sein.',
        'javax.faces.converter.BigIntegerConverter.BIGINTEGER_detail': '{2}: \'{0}\' muss eine Zahl mit einer oder mehr Ziffern sein. Beispiel: {1}',
        'javax.faces.converter.ByteConverter.BYTE': '{2}: \'{0}\' muss eine Zahl zwischen 0 und 255 sein.',
        'javax.faces.converter.ByteConverter.BYTE_detail': '{2}: \'{0}\' muss eine Zahl zwischen 0 und 255 sein. Beispiel: {1}',
        'javax.faces.converter.CharacterConverter.CHARACTER': '{1}: \'{0}\' muss ein gültiges Zeichen sein.',
        'javax.faces.converter.CharacterConverter.CHARACTER_detail': '{1}: \'{0}\' muss ein gültiges ASCII-Zeichen sein.',
        'javax.faces.converter.ShortConverter.SHORT': '{2}: \'{0}\' muss eine Zahl mit einer oder mehr Ziffern sein',
        'javax.faces.converter.ShortConverter.SHORT_detail': '{2}: \'{0}\' muss eine Zahl zwischen -32768 und 32767 sein. Beispiel: {1}',
        'javax.faces.converter.BooleanConverter.BOOLEAN': '{1}: \'{0}\' muss \'wahr\' oder \'falsch\' sein.',
        'javax.faces.converter.BooleanConverter.BOOLEAN_detail': '{1}: \'{0}\' muss \'wahr\' oder \'falsch\' sein.  Ein Wert der nicht \'wahr\' ist wird zu \'falsch\'.',
        'javax.faces.validator.LongRangeValidator.MAXIMUM': '{1}: Der Wert ist größer als das Maximum von \'{0}\'.',
        'javax.faces.validator.LongRangeValidator.MINIMUM': '{1}: Der Wert ist kleiner als das Minimum von \'{0}\'.',
        'javax.faces.validator.LongRangeValidator.NOT_IN_RANGE': '{2}: Wert ist nicht innerhalb der Grenzen von {0} und {1}.',
        'javax.faces.validator.LongRangeValidator.TYPE={0}': 'Wert entsprecht nicht dem vorgegebenen Typ.',
        'javax.faces.validator.DoubleRangeValidator.MAXIMUM': '{1}: Der Wert ist größer als das erlaubte Maximum von \'{0}\'.',
        'javax.faces.validator.DoubleRangeValidator.MINIMUM': '{1}: Der Wert ist kleiner als das erlaubte Minimum von \'{0}\'.',
        'javax.faces.validator.DoubleRangeValidator.NOT_IN_RANGE': '{2}: Attribute nicht innerhalb der Grenzen von {0} und {1}.',
        'javax.faces.validator.DoubleRangeValidator.TYPE={0}': 'Der Wert entspricht nicht dem erwarteten Typ.',
        'javax.faces.converter.FloatConverter.FLOAT': '{2}: \'{0}\' muss eine Zahl mit einer oder mehr Ziffern sein.',
        'javax.faces.converter.FloatConverter.FLOAT_detail': '{2}: \'{0}\' muss eine Zahl zwischen 1.4E-45 und 3.4028235E38 sein. Beispiel: {1}',
        'javax.faces.converter.DateTimeConverter.DATE': '{2}: \'{0}\' kann nicht als Datum interpretiert werden.',
        'javax.faces.converter.DateTimeConverter.DATE_detail': '{2}: \'{0}\' kann nicht als Datum interpretiert werden. Beispiel: {1}',
        'javax.faces.converter.DateTimeConverter.TIME': '{2}: \'{0}\' kann nicht als Zeit interpretiert werden.',
        'javax.faces.converter.DateTimeConverter.TIME_detail': '{2}: \'{0}\' kann nicht als Zeit interpretiert werden. Beispiel: {1}',
        'javax.faces.converter.DateTimeConverter.DATETIME': '{2}: \'{0}\' kann nicht als Datum oder Zeit interpretiert werden.',
        'javax.faces.converter.DateTimeConverter.DATETIME_detail': '{2}: \'{0}\' kann nicht als Datum oder Zeit interpretiert werden. Beispiel: {1}',
        'javax.faces.converter.DateTimeConverter.PATTERN_TYPE': '{1}: Ein \'pattern\' der \'type\' Attribute muss angegeben werden um den Wert \'{0}\' zu konvertieren.', 
        'javax.faces.converter.NumberConverter.CURRENCY': '{2}: \'{0}\' kann nicht als Währung interpretiert werden.',
        'javax.faces.converter.NumberConverter.CURRENCY_detail': '{2}: \'{0}\' kann nicht als Währung interpretiert werden. Beispiel: {1}',
        'javax.faces.converter.NumberConverter.PERCENT': '{2}: \'{0}\' kann nicht als Prozent interpretiert werden.',
        'javax.faces.converter.NumberConverter.PERCENT_detail': '{2}: \'{0}\' kann nicht als Prozent interpretiert werden. Beispiel: {1}',
        'javax.faces.converter.NumberConverter.NUMBER': '{2}: \'{0}\' kann nicht als Zahl interpretiert werden.',
        'javax.faces.converter.NumberConverter.NUMBER_detail': '{2}: \'{0}\' kann nicht als Zahl interpretiert werden. Beispiel: {1}',
        'javax.faces.converter.NumberConverter.PATTERN': '{2}: \'{0}\' kann nicht als Muster interpretiert werden.',
        'javax.faces.converter.NumberConverter.PATTERN_detail': '{2}: \'{0}\' kann nicht als Muster interpretiert werden. Beispiel: {1}',
        'javax.faces.validator.LengthValidator.MINIMUM': '{1}: Die Länge ist kleiner als das Minimum von \'{0}\'.',
        'javax.faces.validator.LengthValidator.MAXIMUM': '{1}: die Länge ist größer als das Maximum von \'{0}\'.',
        'javax.faces.validator.RegexValidator.PATTERN_NOT_SET': 'Ein regulärer Ausdruck darf nicht leer sein.',
        'javax.faces.validator.RegexValidator.PATTERN_NOT_SET_detail': 'Ein regulärer Ausdruck darf nicht leer sein.',
        'javax.faces.validator.RegexValidator.NOT_MATCHED': 'Das Muster stimmt nicht überein.',
        'javax.faces.validator.RegexValidator.NOT_MATCHED_detail': 'Das Muster \'{0}\' des regulären Ausdruck stimmt nicht überein.',
        'javax.faces.validator.RegexValidator.MATCH_EXCEPTION': 'Fehler im regulären Ausdruck.',
        'javax.faces.validator.RegexValidator.MATCH_EXCEPTION_detail': 'Fehler im regulären Ausdruck, \'{0}\'.',
        //optional for bean validation integration in client side validation
        'javax.faces.validator.BeanValidator.MESSAGE': '{0}',
        'javax.validation.constraints.AssertFalse.message': 'Muss falsch sein',
        'javax.validation.constraints.AssertTrue.message': 'Muss wahr sein',
        'javax.validation.constraints.DecimalMax.message': 'Muss kleiner oder gleich {0} sein',
        'javax.validation.constraints.DecimalMin.message': 'Muss größer oder gleich {0} sein',
        'javax.validation.constraints.Digits.message': 'Zahl außerhalb der Grenzen. (<{0} Ziffern>.<{1} Ziffern> möglich)',
        'javax.validation.constraints.Future.message': 'Muss in der Zukunft liegen',
        'javax.validation.constraints.Max.message': 'Muss kleiner oder gleich {0} sein',
        'javax.validation.constraints.Min.message': 'Muss größer oder gleich {0} sein',
        'javax.validation.constraints.NotNull.message': 'darf nicht Null sein',
        'javax.validation.constraints.Null.message': 'Muss Null sein',
        'javax.validation.constraints.Past.message': 'Muss in der Vergangenheit liegen',
        'javax.validation.constraints.Pattern.message': 'Muss gleich "{0}" sein',
        'javax.validation.constraints.Size.message': 'Größe muss zwischen {0} und {1} liegen'
    }
};
```

melloware committed:
Fix #6982: Client side localization script and context-param

melloware commented:
Thanks I updated my PR with your changes!

claudioweiler commented:
Brazilian Portuguese: pt_BR

https://github.com/primefaces/primefaces/wiki/Locales#brazilian-portuguese reviewed

melloware commented:
@claudioweiler Thank you I have updated my PR with your updates!

melloware commented:
Russian updated thanks to Markello.
